ENH: Download zip file from test server

turanoo opened the pull request.

turanoo committed:
ENH: Unzip response and parse the unzipped CSV files

turanoo committed:
ENH: Download zipped file from test server, unzip and write to DB

thermokarst commented on src/config.js:
```diff
@@ -12,6 +12,8 @@ const validateConfigSchema = (config) => {
   return defaultConfigKeys.map(x => configKeys.includes(x)).every(x => x);
 };
 
+// validate config function
```
`// validate config function` -> `// TODO validate config function as part of issue #43`

https://github.com/caporaso-lab/SurveySync/issues/43

thermokarst commented on src/ui.js:
```diff
@@ -13,10 +13,16 @@ function showSurveyConfiguration() {
 function triggerDataUpdate() {
-  if (!verifyDB()) {
-    throw new Error('Please run \'Initialize Survey\' prior to getting data.');
+  // TODO add config validation once we address issue-
```
`// TODO add config validation once we address issue-` -> `// TODO add config validation once we address issue #43`

https://github.com/caporaso-lab/SurveySync/issues/43

thermokarst commented on src/data.js:
```diff
@@ -1,27 +1,29 @@
-function writeDataToDB(table) {
-  const DB = getDB();
+function writeDataToDB(csvName, table) {
+  protectDB(insertDB(csvName));
```
If the database already exists, this will blow up:

![screen shot 2018-05-29 at 5 35 10 pm](https://user-images.githubusercontent.com/274668/40692429-a7dabc26-6366-11e8-9365-94651b2ce4fa.png)

Either, we should check for the existence of the DB (maybe in `insertDB`? That function could be renamed `upsertDB` and perform an [upsert operation](https://en.wiktionary.org/wiki/upsert)). Another option would be to delete the DB sheets after we have used that data to update the "main" sheet, but we don't have any of that code in place. What do you think?

turanoo replied:
I think I like the idea of the upsertDB function better. I think as of right now it makes more sense.

thermokarst commented on src/ui.js:
```diff
@@ -13,10 +13,16 @@ function showSurveyConfiguration() {
+  // TODO add config validation once we address issue-
+  const config = getConfig();
+  const resp = fetchData(config.surveyUrl);
+  const blobs = unzipResponse(resp);
```
I am seeing the following error crop up when the code hits this point:

![screen shot 2018-05-29 at 5 37 58 pm](https://user-images.githubusercontent.com/274668/40692576-799fac6c-6367-11e8-9a78-44df944f445a.png)

The stackdriver log:

```
Incorrect range width, was 1 but should be 3 at writeDataToDB(addon.bundle:6814) at triggerDataUpdate(addon.bundle:6767)
```

I can help you look into this a bit more on Wednesday. This might just be related to the ZIP file I am serving - I used macOS to create it, and sometimes that util is poorly behaved.

thermokarst reviewed (changes requested):
Lookin' great @turanoo! A few minor requests, inline! 🎉

turanoo committed:
SQUASH: Download Zip file & add testing functionality

- add separator between testing functions and main functions in buildMenu
- add function to delete all database sheets
- fix for loop to iterate through csv and import data

thermokarst commented on src/config.js:
```diff
@@ -12,6 +12,17 @@ const validateConfigSchema = (config) => {
 };
 
+// validate config function
+function validateConfig(config) {
```
Looks good. Could be converted to a `.map` and an [`.every`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every) - your call either way!

thermokarst commented on src/testing.js:
```diff
@@ -2,6 +2,19 @@ import { updateConfig } from './config';
 const testingClearExistingConfig = () => updateConfig({});
 
-export default testingClearExistingConfig;
+function deleteDatabaseSheets() {
```
I think we decided to prefix all of the functions in this module with the word "testing" to differentiate them in the global namespace. So, this would look like `testingDeleteDatabaseSheets`.

thermokarst commented on src/testing.js:
```diff
@@ -2,6 +2,19 @@ import { updateConfig } from './config';
+  const sheets = SpreadsheetApp.getActiveSpreadsheet().getSheets();
+  for (let i = 0; i < sheets.length; i += 1) {
+    switch (sheets[i].getSheetName()) {
+      case 'Sheet 1':
```
We probably don't want to hard-code this in here - maybe it makes more sense to invert the data here. We can save all of the DB sheet names in the propertiesservice store when we fetch data. Then, in this function we can look those sheets up from the props and remove them. Another option would be to see if the sheets API allows us to save arbitrary flags to sheet object --- you could mark these somehow as "temporary" sheets. Let me know what you think!

thermokarst commented on src/ui.js:
```diff
@@ -13,17 +13,27 @@ function showSurveyConfiguration() {
+    }
   }
-  writeDataToDB(parseData(fetchData()));
+  SpreadsheetApp.getUi().alert('The Survey Configuration was left empty! Please reinitialize.');
```
How about we throw an `Error` here, instead? That way the UX and UI is consistent.

thermokarst commented on src/ui.js:
```diff
@@ -13,17 +13,27 @@ function showSurveyConfiguration() {
   SpreadsheetApp.getUi().createAddonMenu()
     .addItem('Get Data', 'triggerDataUpdate')
     .addItem('Initialize Survey', 'showSurveyConfiguration')
     .addItem('Test Clearing Existing Configuration', 'testingClearExistingConfig')
```
Can you add a `.addSeparator()` between your testing funcs and the normal funcs? Just to visually partition them.

thermokarst commented on src/data.js:
```diff
@@ -1,27 +1,28 @@
-function writeDataToDB(table) {
-  const DB = getDB();
+function writeDataToDB(csvName, table) {
+  upsertDB(csvName);
```
since this function (`upsertDB`) returns a reference to the DB sheet, you can just assign `DB` here, eliminating the need for the next line!

thermokarst reviewed (changes requested):
Lookin' good @turanoo! A few minor comments inline.

thermokarst reviewed (approved):
LGTM, thanks @turanoo!!
